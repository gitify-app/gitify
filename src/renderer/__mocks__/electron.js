const { namespacedEvent } = require('../../shared/events');

// @ts-ignore
window.Notification = function (title) {
  this.title = title;

  return {
    onclick: vi.fn(),
  };
};

// @ts-ignore
window.Audio = class Audio {
  constructor(path) {
    this.path = path;
  }

  play() {}
};

// @ts-ignore
window.localStorage = {
  store: {},
  getItem: function (key) {
    return this.store[key];
  },
  setItem: function (key, item) {
    this.store[key] = item;
  },
  removeItem: vi.fn(),
};

window.alert = vi.fn();

module.exports = {
  ipcRenderer: {
    send: vi.fn(),
    on: vi.fn(),
    sendSync: vi.fn(),
    invoke: vi.fn((channel, ..._args) => {
      switch (channel) {
        case 'get-platform':
          return Promise.resolve('darwin');
        case namespacedEvent('version'):
          return Promise.resolve('0.0.1');
        case namespacedEvent('safe-storage-encrypt'):
          return Promise.resolve('encrypted');
        case namespacedEvent('safe-storage-decrypt'):
          return Promise.resolve('decrypted');
        default:
          return Promise.reject(new Error(`Unknown channel: ${channel}`));
      }
    }),
  },
  shell: {
    openExternal: vi.fn(),
  },
  webFrame: {
    setZoomLevel: vi.fn(),
    getZoomLevel: vi.fn(),
  },
};
