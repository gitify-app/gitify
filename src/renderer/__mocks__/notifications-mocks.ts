import type { AccountNotifications } from '../types';
import {
  mockEnterpriseNotifications,
  mockGitHubNotifications,
} from '../utils/api/__mocks__/response-mocks';
import {
  mockGitHubCloudAccount,
  mockGitHubEnterpriseServerAccount,
} from './state-mocks';

export const mockAccountNotifications: AccountNotifications[] = [
  {
    account: mockGitHubCloudAccount,
    notifications: mockGitHubNotifications,
    error: null,
  },
  {
    account: mockGitHubEnterpriseServerAccount,
    notifications: mockEnterpriseNotifications,
    error: null,
  },
];

export const mockSingleAccountNotifications: AccountNotifications[] = [
  {
    account: mockGitHubCloudAccount,
    notifications: [mockGitHubNotifications[0]],
    error: null,
  },
];
