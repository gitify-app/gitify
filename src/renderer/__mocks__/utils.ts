import * as helpers from '../utils/helpers';

export function mockDirectoryPath() {
  jest.spyOn(helpers, 'getDirectoryPath').mockReturnValue('/mocked/dir/name');
}

/**
 * Ensure stable snapshots for our randomized emoji use-cases
 */
export function ensureStableEmojis() {
  global.Math.random = jest.fn(() => 0.1);
}
