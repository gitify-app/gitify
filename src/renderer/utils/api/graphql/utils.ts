export function formatAsGitHubSearchSyntax(
  repo: string,
  title: string,
): string {
  return `${title} in:title repo:${repo}`;
}
