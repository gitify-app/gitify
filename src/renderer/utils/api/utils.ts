import type { AxiosResponse } from 'axios';

import type { Hostname } from '../../types';
import { Constants } from '../constants';
import { isEnterpriseServerHost } from '../helpers';

export function getGitHubAPIBaseUrl(hostname: Hostname): URL {
  const url = new URL(Constants.GITHUB_API_BASE_URL);

  if (isEnterpriseServerHost(hostname)) {
    url.hostname = hostname;
    url.pathname = '/api/v3/';
  }
  return url;
}

export function getGitHubGraphQLUrl(hostname: Hostname): URL {
  const url = new URL(Constants.GITHUB_API_GRAPHQL_URL);

  if (isEnterpriseServerHost(hostname)) {
    url.hostname = hostname;
    url.pathname = '/api/graphql';
  }

  return url;
}

export function getNextURLFromLinkHeader(
  response: AxiosResponse,
): string | null {
  const linkHeader = response.headers.link;
  const matches = linkHeader?.match(/<([^<>]+)>;\s*rel="next"/);
  return matches ? matches[1] : null;
}

/**
 * Return true if the request should be made with no-cache
 *
 * @param url
 * @returns boolean
 */
export function shouldRequestWithNoCache(url: string): boolean {
  const parsedUrl = new URL(url);

  switch (parsedUrl.pathname) {
    case '/notifications':
    case '/api/v3/notifications':
      return true;
    default:
      return false;
  }
}
