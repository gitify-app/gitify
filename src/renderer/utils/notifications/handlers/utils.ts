import type { SubjectUser, User } from '../../../typesGitHub';

/**
 * Construct the notification subject user based on an order prioritized list of users
 * @param users array of users in order or priority
 * @returns the subject user
 */
export function getSubjectUser(users: User[]): SubjectUser {
  let subjectUser: SubjectUser = null;

  for (const user of users) {
    if (user) {
      subjectUser = {
        login: user.login,
        html_url: user.html_url,
        avatar_url: user.avatar_url,
        type: user.type,
      };

      return subjectUser;
    }
  }

  return subjectUser;
}
