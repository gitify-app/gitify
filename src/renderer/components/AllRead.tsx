import { type FC, useContext, useMemo } from 'react';

import { AppContext } from '../context/App';
import { Constants } from '../utils/constants';
import { hasAnyFiltersSet } from '../utils/filters';
import { EmojiSplash } from './layout/EmojiSplash';

interface IAllRead {
  fullHeight?: boolean;
}

export const AllRead: FC<IAllRead> = ({ fullHeight = true }: IAllRead) => {
  const { settings } = useContext(AppContext);

  const hasFilters = hasAnyFiltersSet(settings);

  const emoji = useMemo(
    () =>
      Constants.ALL_READ_EMOJIS[
        Math.floor(Math.random() * Constants.ALL_READ_EMOJIS.length)
      ],
    [],
  );

  const heading = `No new ${hasFilters ? 'filtered ' : ''} notifications`;

  return (
    <EmojiSplash emoji={emoji} heading={heading} fullHeight={fullHeight} />
  );
};
