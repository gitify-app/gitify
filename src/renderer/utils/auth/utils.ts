import { BrowserWindow } from '@electron/remote';
import { format } from 'date-fns';
import log from 'electron-log';
import semver from 'semver';
import type {
  Account,
  AuthCode,
  AuthState,
  ClientID,
  GitifyUser,
  Hostname,
  Link,
  Token,
} from '../../types';
import type { UserDetails } from '../../typesGitHub';
import { getAuthenticatedUser } from '../api/client';
import { apiRequest } from '../api/request';
import { Constants } from '../constants';
import { getPlatformFromHostname } from '../helpers';
import type { AuthMethod, AuthResponse, AuthTokenResponse } from './types';

// TODO - Refactor our OAuth2 flow to use system browser and local app gitify://callback - see #485 #561 #654
export function authGitHub(
  authOptions = Constants.DEFAULT_AUTH_OPTIONS,
): Promise<AuthResponse> {
  return new Promise((resolve, reject) => {
    // Build the OAuth consent page URL
    const authWindow = new BrowserWindow({
      width: 548,
      height: 736,
      show: true,
    });

    const authUrl = new URL(`https://${authOptions.hostname}`);
    authUrl.pathname = '/login/oauth/authorize';
    authUrl.searchParams.append('client_id', authOptions.clientId);
    authUrl.searchParams.append('scope', Constants.AUTH_SCOPE.toString());

    const session = authWindow.webContents.session;
    session.clearStorageData();

    authWindow.loadURL(authUrl.toString());

    const handleCallback = (url: Link) => {
      const raw_code = /code=([^&]*)/.exec(url) || null;
      const authCode =
        raw_code && raw_code.length > 1 ? (raw_code[1] as AuthCode) : null;
      const error = /\?error=(.+)$/.exec(url);
      if (authCode || error) {
        // Close the browser if code found or error
        authWindow.destroy();
      }
      // If there is a code, proceed to get token from github
      if (authCode) {
        resolve({ authCode, authOptions });
      } else if (error) {
        reject(
          "Oops! Something went wrong and we couldn't " +
            'log you in using GitHub. Please try again.',
        );
      }
    };

    // If "Done" button is pressed, hide "Loading"
    authWindow.on('close', () => {
      authWindow.destroy();
    });

    authWindow.webContents.on(
      'did-fail-load',
      (_event, _errorCode, _errorDescription, validatedURL) => {
        if (validatedURL.includes(authOptions.hostname)) {
          authWindow.destroy();
          reject(
            `Invalid Hostname. Could not load https://${authOptions.hostname}/.`,
          );
        }
      },
    );

    authWindow.webContents.on('will-redirect', (event, url) => {
      event.preventDefault();
      handleCallback(url as Link);
    });

    authWindow.webContents.on('will-navigate', (event, url) => {
      event.preventDefault();
      handleCallback(url as Link);
    });
  });
}

export async function getUserData(
  token: Token,
  hostname: Hostname,
): Promise<GitifyUser> {
  const response: UserDetails = (await getAuthenticatedUser(hostname, token))
    .data;

  return {
    id: response.id,
    login: response.login,
    name: response.name,
    avatar: response.avatar_url,
  };
}

export async function getToken(
  authCode: AuthCode,
  authOptions = Constants.DEFAULT_AUTH_OPTIONS,
): Promise<AuthTokenResponse> {
  const url =
    `https://${authOptions.hostname}/login/oauth/access_token` as Link;
  const data = {
    client_id: authOptions.clientId,
    client_secret: authOptions.clientSecret,
    code: authCode,
  };

  const response = await apiRequest(url, 'POST', data);
  return {
    hostname: authOptions.hostname,
    token: response.data.access_token,
  };
}

export async function addAccount(
  auth: AuthState,
  method: AuthMethod,
  token: Token,
  hostname: Hostname,
): Promise<AuthState> {
  let newAccount = {
    hostname: hostname,
    method: method,
    platform: getPlatformFromHostname(hostname),
    token: token,
  } as Account;

  newAccount = await refreshAccount(newAccount);

  return {
    accounts: [...auth.accounts, newAccount],
  };
}

export function removeAccount(auth: AuthState, account: Account): AuthState {
  const updatedAccounts = auth.accounts.filter(
    (a) => a.token !== account.token,
  );

  return {
    accounts: updatedAccounts,
  };
}

export async function refreshAccount(account: Account): Promise<Account> {
  try {
    const res = await getAuthenticatedUser(account.hostname, account.token);

    // Refresh user data
    account.user = {
      id: res.data.id,
      login: res.data.login,
      name: res.data.name,
      avatar: res.data.avatar_url,
    };

    account.version = extractHostVersion(
      res.headers['x-github-enterprise-version'],
    );
  } catch (err) {
    log.error(
      '[refreshAccount]: failed to refresh account for user',
      account.user.login,
      err,
    );
  }

  return account;
}

export function extractHostVersion(version: string | null): string {
  if (version) {
    return semver.valid(semver.coerce(version));
  }

  return 'latest';
}

export function getDeveloperSettingsURL(account: Account): Link {
  const settingsURL = new URL(`https://${account.hostname}`);

  switch (account.method) {
    case 'GitHub App':
      settingsURL.pathname = '/settings/apps';
      break;
    case 'OAuth App':
      settingsURL.pathname = '/settings/developers';
      break;
    case 'Personal Access Token':
      settingsURL.pathname = '/settings/tokens';
      break;
    default:
      settingsURL.pathname = '/settings';
      break;
  }
  return settingsURL.toString() as Link;
}

export function getNewTokenURL(hostname: Hostname): Link {
  const date = format(new Date(), 'PP p');
  const newTokenURL = new URL(`https://${hostname}/settings/tokens/new`);
  newTokenURL.searchParams.append('description', `Gitify (Created on ${date})`);
  newTokenURL.searchParams.append('scopes', Constants.AUTH_SCOPE.join(','));

  return newTokenURL.toString() as Link;
}

export function getNewOAuthAppURL(hostname: Hostname): Link {
  const date = format(new Date(), 'PP p');
  const newOAuthAppURL = new URL(
    `https://${hostname}/settings/applications/new`,
  );
  newOAuthAppURL.searchParams.append(
    'oauth_application[name]',
    `Gitify (Created on ${date})`,
  );
  newOAuthAppURL.searchParams.append(
    'oauth_application[url]',
    'https://www.gitify.io',
  );
  newOAuthAppURL.searchParams.append(
    'oauth_application[callback_url]',
    'https://www.gitify.io/callback',
  );

  return newOAuthAppURL.toString() as Link;
}

export function isValidHostname(hostname: Hostname) {
  return /^([a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,6}$/i.test(
    hostname,
  );
}

export function isValidClientId(clientId: ClientID) {
  return /^[A-Z0-9_]{20}$/i.test(clientId);
}

export function isValidToken(token: Token) {
  return /^[A-Z0-9_]{40}$/i.test(token);
}

export function getAccountUUID(account: Account): string {
  return btoa(`${account.hostname}-${account.user.id}-${account.method}`);
}

export function hasAccounts(auth: AuthState) {
  return auth.accounts.length > 0;
}

export function hasMultipleAccounts(auth: AuthState) {
  return auth.accounts.length > 1;
}
