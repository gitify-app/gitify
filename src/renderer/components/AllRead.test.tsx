import { render } from '@testing-library/react';
import { ensureStableEmojis, mockDirectoryPath } from '../__mocks__/utils';
import { AllRead } from './AllRead';

describe('renderer/components/AllRead.tsx', () => {
  beforeEach(() => {
    ensureStableEmojis();
    mockDirectoryPath();
  });

  it('should render itself & its children', () => {
    const tree = render(<AllRead />);

    expect(tree).toMatchSnapshot();
  });
});
