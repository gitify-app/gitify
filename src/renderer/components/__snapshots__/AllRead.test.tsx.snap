// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer/components/AllRead.tsx should render itself & its children - no filters 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Stack__StyledStack-sc-x3xa2i-0 duhwY h-full"
        data-align="center"
        data-direction="vertical"
        data-justify="center"
        data-padding="spacious"
        data-wrap="nowrap"
      >
        <div
          class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
          data-align="center"
          data-direction="vertical"
          data-gap="condensed"
          data-justify="center"
          data-padding="none"
          data-wrap="nowrap"
        >
          <div
            class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
            data-align="center"
            data-direction="vertical"
            data-gap="spacious"
            data-justify="start"
            data-padding="none"
            data-wrap="nowrap"
          >
            <div
              class="Box-sc-g0xbh4-0 text-5xl"
            >
              <img
                alt="ðŸŽŠ"
                class="emoji"
                draggable="false"
                src="images/twemoji/1f38a.svg"
              />
            </div>
            <div
              class="Box-sc-g0xbh4-0 text-xl font-semibold"
            >
              No new  notifications
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Stack__StyledStack-sc-x3xa2i-0 duhwY h-full"
      data-align="center"
      data-direction="vertical"
      data-justify="center"
      data-padding="spacious"
      data-wrap="nowrap"
    >
      <div
        class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
        data-align="center"
        data-direction="vertical"
        data-gap="condensed"
        data-justify="center"
        data-padding="none"
        data-wrap="nowrap"
      >
        <div
          class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
          data-align="center"
          data-direction="vertical"
          data-gap="spacious"
          data-justify="start"
          data-padding="none"
          data-wrap="nowrap"
        >
          <div
            class="Box-sc-g0xbh4-0 text-5xl"
          >
            <img
              alt="ðŸŽŠ"
              class="emoji"
              draggable="false"
              src="images/twemoji/1f38a.svg"
            />
          </div>
          <div
            class="Box-sc-g0xbh4-0 text-xl font-semibold"
          >
            No new  notifications
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`renderer/components/AllRead.tsx should render itself & its children - with filters 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Stack__StyledStack-sc-x3xa2i-0 duhwY h-full"
        data-align="center"
        data-direction="vertical"
        data-justify="center"
        data-padding="spacious"
        data-wrap="nowrap"
      >
        <div
          class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
          data-align="center"
          data-direction="vertical"
          data-gap="condensed"
          data-justify="center"
          data-padding="none"
          data-wrap="nowrap"
        >
          <div
            class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
            data-align="center"
            data-direction="vertical"
            data-gap="spacious"
            data-justify="start"
            data-padding="none"
            data-wrap="nowrap"
          >
            <div
              class="Box-sc-g0xbh4-0 text-5xl"
            >
              <img
                alt="ðŸŽŠ"
                class="emoji"
                draggable="false"
                src="images/twemoji/1f38a.svg"
              />
            </div>
            <div
              class="Box-sc-g0xbh4-0 text-xl font-semibold"
            >
              No new filtered  notifications
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Stack__StyledStack-sc-x3xa2i-0 duhwY h-full"
      data-align="center"
      data-direction="vertical"
      data-justify="center"
      data-padding="spacious"
      data-wrap="nowrap"
    >
      <div
        class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
        data-align="center"
        data-direction="vertical"
        data-gap="condensed"
        data-justify="center"
        data-padding="none"
        data-wrap="nowrap"
      >
        <div
          class="Stack__StyledStack-sc-x3xa2i-0 duhwY"
          data-align="center"
          data-direction="vertical"
          data-gap="spacious"
          data-justify="start"
          data-padding="none"
          data-wrap="nowrap"
        >
          <div
            class="Box-sc-g0xbh4-0 text-5xl"
          >
            <img
              alt="ðŸŽŠ"
              class="emoji"
              draggable="false"
              src="images/twemoji/1f38a.svg"
            />
          </div>
          <div
            class="Box-sc-g0xbh4-0 text-xl font-semibold"
          >
            No new filtered  notifications
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
